Support environment variable in BUILD.yaml + having tag in build.tag

skovhus opened the pull request:
This PR updates Brick with two new features:
- Support environment variable expansion in BUILD.yaml (limited to environment variables prefixed with `BRICK_`) 
- Support having Docker tags in `build.tag` (that currently just supports having a repository)

The primary use case is to inject commit SHA into build steps on CI. It means that images can be tagged with a unique value instead of always being tagged with `latest`.

### Example

```
steps:

  build:
    inputs:
      - metadata
    tag: eu.gcr.io/server:${BRICK_DRONE_COMMIT_SHA:-master}

  deploy:
    commands:
      - gcloud run deploy NAME --image eu.gcr.io/server:${BRICK_DRONE_COMMIT_SHA:-master}

```

skovhus committed:
Support environment variable expansion in BUILD.yaml

Use case: inject commit SHA into build steps on CI

madsnedergaard commented on brick/lib.py:
```diff
@@ -111,6 +111,8 @@ def get_config(target):
             # TODO: we could be basic sanity checking here
-            return yaml.load(f, Loader=yaml.FullLoader)
+            data = os.path.expandvars(f.read())
+            print(data)
```
Should this print still be there?

madsnedergaard reviewed (commented):
Nice! Should we also update the examples in README with this information?

madsnedergaard commented:
I see now that the example in README doesn't really cover any of this, so please ignore that suggestion 😄

skovhus committed:
Allow build.tag to actually contain a tag

in case there is not image name (`steps.build.tag`):
['src_backend_hasura_build:latest', 'src_backend_hasura_build:ks-bloom-master-tags', 'src_backend_hasura:latest', 'src_backend_hasura:ks-bloom-master-tags']

in case there is a image name (`steps.build.tag`) with a tag:
['src_backend_hasura_build:latest', 'src_backend_hasura_build:ks-bloom-master-tags', 'eu.gcr.io/tmrowapp/graphql-server:1']

in case there is a image name (`steps.build.tag`) without a tag:
['src_backend_hasura_build:latest', 'src_backend_hasura_build:ks-bloom-master-tags', 'eu.gcr.io/tmrowapp/graphql-server:latest', 'eu.gcr.io/tmrowapp/graphql-server:ks-bloom-master-tags']

madsnedergaard commented:
For my own understanding, is the following assumptions correct? :)

## Case 1: no image name

```yaml
  build:
    inputs:
      - ...
```

Will create:
```
['src_server:latest', 'src_server:my-branch-name']
```

## Case 2: image name with tag

```yaml
  build:
    inputs:
      - ...
    tag: eu.gcr.io/server:production
```

Will create:
```
['src_server:latest', 'src_server:my-branch-name', 'eu.gcr.io/server:production']
```


## Case 3: image name without tag

```yaml
  build:
    inputs:
      - ...
    tag: eu.gcr.io/server
```

Will create:
```
['src_server:latest', 'src_server:my-branch-name', 'eu.gcr.io/server:latest', 'eu.gcr.io/server:my-branch-name']

skovhus commented:
I've explained the tagging in the description of this commit  https://github.com/tmrowco/brick/pull/64/commits/faced4b2c16f3360f6315f12f1c43dec3c82e772 :)

madsnedergaard commented:
> I've explained the tagging in the description of this commit [faced4b](https://github.com/tmrowco/brick/commit/faced4b2c16f3360f6315f12f1c43dec3c82e772) :)

Yeah I saw that, I just wanted to flesh out the example to be sure it was correct :)

madsnedergaard commented:
And the purpose of this change (for our use case) would be do to something like this, right?

```yaml
  build:
    inputs:
      - ...
    tag: eu.gcr.io/server:$COMMIT_SHA

  deploy:
    commands:
      - gcloud run deploy NAME --image eu.gcr.io/server:$COMMIT_SHA
```

skovhus commented:
> Would it make more sense to use the dependencies-hash for tagging?

The dependency hash is not nice for traceability (i.e. the purpose of the git hash is to be able to route it back to a specific commit)...

> Would this not force a new deployment of every single package despite no actual changes? Or would it just add the new tag ($COMMIT_SHA) to the existing image? 🤔

That is true. 🤔 But isn't that the same behaviour as our other packages?

> Case 2: image name with tag

Would only push ['eu.gcr.io/server:production']

> Case 3: image name without tag

Would  only push ['eu.gcr.io/server:latest', 'eu.gcr.io/server:my-branch-name']

So in general the repository that we extract from the folder structure/name is gone when using `build.tag`. That is also the current behaviour.

madsnedergaard commented:
> > Would it make more sense to use the dependencies-hash for tagging?
> 
> The dependency hash is not nice for traceability (i.e. the purpose of the git hash is to be able to route it back to a specific commit)...
> 
> > Would this not force a new deployment of every single package despite no actual changes? Or would it just add the new tag ($COMMIT_SHA) to the existing image? 🤔
> 
> That is true. 🤔 But isn't that the same behaviour as our other packages?

I'm unsure if this means deployment in CI will take way longer, but I guess the easiest way of finding out is to merge this and test it ¯\\\_(ツ)\_/¯

> > Case 2: image name with tag
> 
> Would only push ['eu.gcr.io/server:production']
> 
> > Case 3: image name without tag
> 
> Would only push ['eu.gcr.io/server:latest', 'eu.gcr.io/server:my-branch-name']
> 
> So in general the repository that we extract from the folder structure/name is gone when using `build.tag`. That is also the current behaviour.

Ah okay, didn't realise that. Makes sense 👍

madsnedergaard reviewed (approved):
Approving this so we can test the impact when used in CI :)

skovhus committed:
Support bash style default values for environment variables

And ensure they start with BRICK (for security and transparency)

skovhus committed:
Document that expand_brick_environment_variables works

skovhus commented:
I've changed the implementation a bit so:
- not all environment variables are replaced (this is a security risk and also confusing). Instead we only replace environment variables that is prefixed with `BRICK_` (similar to how Create React app is using `REACT_APP_`).
- as the `build.tag` is used for local development, then I updated the code to support bash style defaults `${BRICK_FOO:-default}`

I rolled my own little regular expression for this as the two requirements was not covered by any third party that I could find. I've documented it with some unit tests, so it should be safe.

madsnedergaard commented on brick/lib.py:
```diff
@@ -107,10 +108,47 @@ def get_relative_config_path(target):
     return os.path.relpath(get_config_path(target), start=ROOT_PATH)
 
 
+def expand_brick_environment_variables(before_expansion: str) -> str:
```
Seeing as this is now becoming quite specific and it's hard to read/understand regex, would you mind adding a comment here (or somewhere else in the code) about about usage and include an example of how this can be used? :)

**EDIT:** Just saw the test now which does exactly this, but maybe we could add a comment pointing to the test for examples of usage?

skovhus replied:
https://github.com/tmrowco/brick/pull/64/commits/6bc22a7a403af31e039817cf900646bcf91391c7

skovhus committed:
Test get_build_repository_and_tag

madsnedergaard commented on tests/conftest.py:
```diff
@@ -0,0 +1,15 @@
+import os
```
Would have preferred to see the test-setup part in a separate PR, but I guess it was blocking for adding the tests ;)

skovhus replied:
I can pull it out. :)

madsnedergaard replied:
Nah, it looks good to me :)

skovhus replied:
https://github.com/tmrowco/brick/pull/65 :)

skovhus committed:
Support numbers in environment variables

skovhus committed:
Add documention for expand_brick_environment_variables
